pkg/duck/listable.go

```diff
@@ -0,0 +1,165 @@
+		informerFactory: &duck.TypedInformerFactory{
+			Client:       dynamicclient.Get(ctx),
+			Type:         listable,
+			ResyncPeriod: controller.GetResyncPeriod(ctx),
+			StopChannel:  ctx.Done(),
+		},
+		tracker:  tracker.New(callback, lease),
+		concrete: map[schema.GroupVersionResource]informerListerPair{},
+	}
+}
+
+type listableTracker struct {
+	informerFactory duck.InformerFactory
+	tracker         tracker.Interface
+
+	concrete     map[schema.GroupVersionResource]informerListerPair
+	concreteLock sync.RWMutex
+}
+
+type informerListerPair struct {
+	informer cache.SharedIndexInformer
+	lister   cache.GenericLister
+}
+
+// ensureTracking ensures that there is an informer watching and sending events to tracker for the
+// concrete GVK. It also ensures that there is the corresponding lister for that informer.
+func (t *listableTracker) ensureTracking(ref corev1.ObjectReference) error {
+	if equality.Semantic.DeepEqual(ref, &corev1.ObjectReference{}) {
+		return errors.New("cannot track empty object ref")
+	}
+	gvk := ref.GroupVersionKind()
+	gvr, _ := meta.UnsafeGuessKindToResource(gvk)
+
+	t.concreteLock.RLock()
+	_, present := t.concrete[gvr]
+	t.concreteLock.RUnlock()
+	if present {
+		// There is already an informer running for this GVR, we don't need or want to make
+		// a second one.
+		return nil
+	}
+	// Tracking has not been setup for this GVR.
+	t.concreteLock.Lock()
+	defer t.concreteLock.Unlock()
+	// Now that we have acquired the write lock, make sure no one has added tracking handlers.
+	if _, present = t.concrete[gvr]; present {
+		return nil
+	}
+	informer, lister, err := t.informerFactory.Get(gvr)
+	if err != nil {
+		return err
+	}
+	informer.AddEventHandler(controller.HandleAll(
+		// Call the tracker's OnChanged method, but we've seen the objects coming through
+		// this path missing TypeMeta, so ensure it is properly populated.
+		controller.EnsureTypeMeta(
+			t.tracker.OnChanged,
+			gvk,
+		),
+	))
+	t.concrete[gvr] = informerListerPair{informer: informer, lister: lister}
+	return nil
+}
+
+// TrackInNamespace satisfies the ListableTracker interface.
+func (t *listableTracker) TrackInNamespace(obj metav1.Object) Track {
+	return func(ref corev1.ObjectReference) error {
+		// This is often used by Trigger and Subscription, both of which pass in refs that do not
+		// specify the namespace.
+		ref.Namespace = obj.GetNamespace()
+		if err := t.ensureTracking(ref); err != nil {
+			return err
+		}
+		return t.tracker.Track(ref, obj)
+	}
+}
+
+// ListerFor satisfies the ListableTracker interface.
+func (t *listableTracker) ListerFor(ref corev1.ObjectReference) (cache.GenericLister, error) {
+	if equality.Semantic.DeepEqual(ref, &corev1.ObjectReference{}) {
+		return nil, errors.New("cannot get lister for empty object ref")
+	}
+	gvk := ref.GroupVersionKind()
+	gvr, _ := meta.UnsafeGuessKindToResource(gvk)
+
+	t.concreteLock.RLock()
+	informerListerPair, present := t.concrete[gvr]
+	t.concreteLock.RUnlock()
```

Comment:
Why not a standard defer? We don't use this pattern in `ensureTracking` because we may need to acquire the write lock before the method returns.

Both this function and `InformerFor`.

Reply:
Good point, done, thanks!